packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ImagePickerDelegate.java

```diff
@@ -0,0 +1,254 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugins.imagepicker;
+
+import android.Manifest;
+import android.app.Activity;
+import android.content.Intent;
+import android.content.pm.PackageManager;
+import android.content.pm.ResolveInfo;
+import android.media.MediaScannerConnection;
+import android.net.Uri;
+import android.provider.MediaStore;
+import android.support.v4.app.ActivityCompat;
+import android.support.v4.content.FileProvider;
+import io.flutter.plugin.common.MethodCall;
+import io.flutter.plugin.common.MethodChannel;
+import io.flutter.plugin.common.PluginRegistry;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.UUID;
+
+public class ImagePickerDelegate
+    implements PluginRegistry.ActivityResultListener,
+        PluginRegistry.RequestPermissionsResultListener {
+  private static final int REQUEST_CODE_CHOOSE_FROM_GALLERY = 2342;
+  private static final int REQUEST_CODE_TAKE_WITH_CAMERA = 2343;
+
+  private static final int REQUEST_EXTERNAL_STORAGE_PERMISSION = 2344;
+  private static final int REQUEST_CAMERA_PERMISSION = 2345;
+
+  private final Activity activity;
+  private final File externalFilesDirectory;
+  private final ImageResizer imageResizer;
+  private final String providerName;
+
+  private Uri pendingCameraImageUri;
+  private MethodChannel.Result pendingResult;
+  private MethodCall methodCall;
+
+  public ImagePickerDelegate(
+      Activity activity, File externalFilesDirectory, ImageResizer imageResizer) {
+    this.activity = activity;
+    this.externalFilesDirectory = externalFilesDirectory;
+    this.imageResizer = imageResizer;
+    this.providerName = activity.getPackageName() + ".flutter.image_provider";
+  }
+
+  public void chooseImageFromGallery(MethodCall methodCall, MethodChannel.Result result) {
+    if (!setPendingMethodCallAndResult(methodCall, result)) {
+      return;
```

Comment:
I would find the code easier to understand if the finishWIthError would happen here and not inside the setPendingMethodCallAndResult. finishWithError is a strange side effect of a setter.